[Doc] Update meta.md (#6452)
taichi-dev/taichi: docs/lang/articles/advanced/meta.md

    return result;
}
```

When this function is called in the program, maybe in different places and operates on different array-like types `T`, the compiler will generate a version of `sum` for each `T`, as long as `T` implements the `length` method to allow you get the array length, and can be intrated over through indices. In other words, with template programming, you only write the same code once and the compiler automatically generates other versions for you.

Taichi has a similar functionality for template programming: By using `ti.template()` as an argument type hint, you can pass any Python object that Taichi's JIT compiler accepts into a kernel (see [../kernels/kernel_function#arguments]).

Let's write a function `sum` (our old friend) to illustrate this. This `sum` function will take in a Taichi field and return the sum of all its entries.

```python {2}
@ti.kernel
def sum(x: ti.template()) -> float:
    result = 0.0
    for i in ti.grouped(x):
        result += x[i]

f1d = ti.field(float, shape=100)
f2d = ti.field(float, shape=(5, 5))
f3d = ti.field(float, shape=(10, 10, 10))
g3d = ti.field(int, shape=(10, 10, 10))

sum(f1d)
sum(f2d)
sum(f3d)
sum(g3d)
```

> As can be seen from the code above, the function `sum` works for fields of any shape and any dtype. There's no need to bother about the shape of the field. This is very handy for physical simulations as the same function can be used in both 2D and 3D scenarios.

Reviewer: ```suggestion

```

> Note the function `ti.group()` is critial to our dimensionality-independent programming: In general, you will need `d` independent indices to loop over a field of dimension `d`, with one index for each axis. Taichi's `ti.grouped` puts the `d` loop indices into a `d`-dimentional integer vector, and the `for` loop is parallelized for all such multidimensional indices.

Reviewer: ```suggestion
The function `ti.group()` is critical to dimensionality-independent programming. Normally, you need `d` independent indices to loop over a field of dimension `d`, with each index for an axis. Taichi's `ti.grouped` puts the `d` loop indices into a `d`-dimensional integer vector, and the `for` loop is parallelized for all multidimensional indices.
```

> We should mention a difference between Taichi and C++ in template metaprogramming: C++ compilers will generate a version of `sum` for each different type `T`; meanwhile Taichi's compiler **recompiles** the kernel each time it finds an argument of a different type is encounted. In the example above, since the fields are of different shapes, or the same shape but of different dtypes, each of the four calls to `sum` will trigger a compilation:

Reviewer: ```suggestion
Here comes a major difference between Taichi and C++ in template metaprogramming: C++ compilers generate a version of `sum` for each different type `T`; while Taichi's compiler **recompiles** the kernel each time it encounters an argument of a different type. In the example above, the fields are of different shapes or different dtypes, and each of the four calls to `sum` triggers a compilation:
```

> We should mention a difference between Taichi and C++ in template metaprogramming: C++ compilers will generate a version of `sum` for each different type `T`; meanwhile Taichi's compiler **recompiles** the kernel each time it finds an argument of a different type is encounted. In the example above, since the fields are of different shapes, or the same shape but of different dtypes, each of the four calls to `sum` will trigger a compilation:

Reviewer: Can we just say "encounters a different argument"?

```python
sum(f1d)  # Compilation
sum(f2d)  # Recompilation
sum(f3d)  # Recompilation
sum(g3d)  # Recompilation
```

:::note
If a template parameter is not a Taichi object, it cannot be reassigned inside Taichi kernel. For example:

```python
x = [1, 2, 3]
@ti.kernel
def error_reassign(x: ti.template()):
    x = ti.math.vec3(1, 2, 3)  # Error!
```

:::

:::note
The template parameters are inlined into the generated kernel after compilation.
:::

> ## Compile-time evaluations

Reviewer: ```suggestion
### Compile-time evaluations
```

> Using compile-time evaluation allows for some computation to be executed when kernels are instantiated. This helps the compiler to conduct optimization and reduce

Reviewer: ```suggestion
Using compile-time evaluations allow for some computation to be executed when kernels are instantiated. This helps the compiler to conduct optimization and reduce
```

computational overhead at runtime. Or in other words, we do more at compile time to save effort for runime.

### Static Scope

> `ti.static` is a function which receives one argument. It is a hint for the compiler to evaluate the argument at compile time.

Reviewer: ```suggestion
The function `ti.static` receives one argument only and serves as a hint for the compiler to evaluate the argument at compile time.
```

  print(3)
```

> ## When to use `ti.static` with for loops

Reviewer: ```suggestion
### When to use `ti.static` with for loops
```

The main reason to use `ti.static` with for loops is to unroll the loop to improve runtime performance (see [Compile-time evaluations](#compile-time-evaluations)).

An unrolling verison is often faster because it reduces branch overhead:

1. It saves the instructions for incrementing the loop counter and checking the end-of-loop condition.

> 2. It provides more instructions for the compiler to schedule across the whole iterations.

Author: No, not schedule iterations, compiler schedule instructions in the iterations

> This performance improvement does not come for free: it also costs longer complie time and generates larger binaries. Especially when the number of iterations is large or the loop body is complex.

Reviewer: ```suggestion
The performance improvement does not come for free but leads to longer compile time and larger binaries, especially when the number of iterations is large or when the loop body is complex.
```

Author: I think the suggested change "... but leads to ..." is not as good as the original one?

Reviewer: I was thinking about reducing the number of verbs... We can revert to the original form with the colon but I would suggest phrasing it in this way: It takes longer compile time and generates larger binaries, especially when...

## Compile-time recursion of `ti.func`

> A compile-time recursive function is a function with recursion that can be recursively inlined at compile time. The condition which determines when to cease recursion is evaluated at compile time, hence must be known to the compiler.

Reviewer: ```suggestion
A compile-time recursive function is a function with recursion that can be recursively inlined at compile time. The condition that determines when to cease recursion is evaluated at compile time, hence must be known to the compiler.
```
